doc/protocol.rst

```diff
@@ -0,0 +1,270 @@
+header. This is an advisory request; the exact frequency of
+checkpoints is up to the discretion of the server. The value is
+the approximate number of bytes to stream before sending a
+checkpoint message.
+
+The client tunes the checkpoint frequency based on network speed
+and reliability. Since it's expected that that checkpoint will
+mainly be used for larger transfers over wide area networks, the
+default value is disabled (0).
+
+The checkpoint object must be encoded in a manner that is valid
+as an HTTP header value and as a chunk extension value. The HTTP
+specification is not entirely clear on valid values.  Using only
+printable ASCII characters, excluding semi-colon, will suffice.
+URL encoding of checkpoint object is recommended.
+
+It is expected that a similar checkpoint/resume facility will
+be used in a write API when it is defined by GA4GH.
+
+Goals and Rationale
+-------------------
+
+The goals for introducing a streaming protocol are:
+
+-  Improve performance by reducing latency of multiple page responses
+-  Providing efficient binary encoding as well as simple JSON encoding
+-  Simplify server implementation
+-  Simplify client programming paradigm
+-  Allow resume larger result transfers that fail
+-  Allow results of mixed object types without creating new containers
+
+Rationale for a streaming protocol
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The original GA4GH paging protocol offers a simple client
+interface that allows clients to read a complete JSON documents
+'off-the-wire'. The returned response objects contained a
+homogeneous vector of results along with a next page object. This
+also allows for the easy resumption of failed transfers.
+
+However, during the implementing of the protocol, drawbacks have
+been recognized:
+
+- Paging introduces latency, as the client must get a complete
+  response and parse the document before it can issue the
+  request for the next page. Large pages make for poor
+  interactive responsiveness, and small pages lead to a high
+  protocol overhead.
+- Paging was performance limiting due to the need to buffer the
+  returned JSON document. It requires a trade-off between
+  client/server memory and the number of requests. Even if a
+  given client can dedicate a lot of memory for a transfer, the
+  server must impose limits to prevent DoS attacks and manage an
+  unpredictable request load.
+- Paging can make the implementation of a server complex. This is due the
+  needs to be able to efficiently resume every query at an arbitrary point
+  determined by the client.  It's easy of implementation varies with the type
+  of data store and complexity of queries.  While many ReST APIs implement
+  paging, it's not clear that specifying in an interoperable API independent
+  of the implementation is desirable.
+
+Checkpoint rationale
+~~~~~~~~~~~~~~~~~~~~
+
+On of the apparent goals of paging is to allow for resumes large
+transfers on failure.  With paging, the error recover is part of
```

Comment:
One of the apparent goals of paging is to allow for resumption of large transfers upon transmission failure.